### DISCUSSION TITLE
CSG Operations

### DISCUSSION THREAD

**Discussion by LouisNoailles** [AcademySoftwareFoundation/openvdb#1118] [Q&A]
### Environment
**Operating System:** Ubuntu 20.04
**Version / Commit SHA:** 0f93ce6aebba74ac237b32ca4f63cf861dfa1d4d
**Other:** compiler: g++, c++ 17

### Describe the bug
I am trying to do csg operations such as union and I am doing this
`openvdb::FloatGrid::ConstPtr
            copyOfGridA = grid->deepCopy(),
            copyOfGridB = vdb->deepCopy();
        
	    openvdb::FloatGrid::Ptr res = openvdb::tools::csgUnionCopy(*grid, *vdb);
        
        // Restore the original grids.
        grid = copyOfGridA->deepCopy();
        vdb = copyOfGridB->deepCopy();
        return res;`

the two grids have the same transform and I do no rotations on them but the final grid is equal to grid 

### Expected behavior
I would expect to have grid U vdb but I get only grid as a result

**Comment by danrbailey** [comment]
I might need a little more sample code to try and guess why you might not be getting the expected result, however I will say that given that you're deep-copying the inputs before calling `openvdb::tools::csgUnionCopy()`, it would probably make sense to just call `openvdb::tools::csgUnion()` which steals data from the second tree and inserts it into the first. In fact, we have been considering removing `openvdb::tools::csgUnionCopy()` as it's not used by any code in the library currently and the new threaded implementation of `openvdb::tools::csgUnion()` is just as fast.

**Reply by LouisNoailles** [reply]
I also tried csgUnion() like this :
`openvdb::FloatGrid::Ptr res = grid->deepCopy();
openvdb::FloatGrid::Ptr copy = vdb->deepCopy();

openvdb::tools::csgUnion(*res, *vdb)
// Restore the original grids.
vdb = copy->deepCopy();
return res;`

but same restult I only get in the union the content of "grid" and not "grid U vdb"

**Reply by LouisNoailles** [reply]
Is meshtoVolume produce a level set as openvdb understand it or to a for volume ? and is there a way to convert a volume t a level set?
